pyirf/interpolators/base_interpolators.py

```diff
@@ -0,0 +1,235 @@
+"""Base classes for interpolators"""
+import numpy as np
+from pyirf.binning import bin_center
+from scipy.spatial import Delaunay, QhullError
+from abc import ABCMeta, abstractmethod
+
+
+class BaseInterpolator(metaclass=ABCMeta):
+    """
+    Base class for all interpolators, only knowing grid-points,
+    providing a common __call__-interface and doing sanity checks.
+    """
+
+    def __init__(self, grid_points):
+        """BaseInterpolator constructor, doing sanity checks and ensuring
+        correct shapes of inputs
+
+        Parameters
+        ----------
+            grid_points: np.ndarray, shape=(n_points, n_dims):
+                Grid points at which interpolation templates exist
+
+        Raises
+        ------
+            TypeError:
+                When grid_points is not a np.ndarray
+            TypeError:
+                When grid_point has dtype object
+            ValueError:
+                When there are too few points in grid_points to span a volume
+                in the grid dimension.
+        """
+        if not isinstance(grid_points, np.ndarray):
+            raise TypeError("Input grid_points is not a numpy array.")
+        if grid_points.dtype == "O":
```

Comment:
if you expect the dtype to be, e.g. floats or similar, you might use [np.can_cast](https://numpy.org/doc/stable/reference/generated/numpy.can_cast.html) to check for it to be what is expected, rather than just checking that is not an object.

Reply:
This is meant to only prevent inputs like 
```
grid_points = np.array([[1, 2], [1, 2, 3]], dtype=object)
```
to be accepted. Thinking about how often this might occur this actually seems quite defensive, but should not hurt to have it.

Reply:
But it just needs to be something numerical for interpolation to work correct? It seems odd to only check for objects, when strings would also be bad (e.g. `np.array([1,2],[1,2,3], dtype='S'` should also fail with a wrong dtype error)

Reply:
Good point, added a check and a test